File: playlist-tool.js
#!/usr/bin/env node
/**
 * Nexus TV OS - Command Line Playlist Extraction Tool
 * Simple CLI that extracts URLs from playlists and creates output files
 */

const fs = require('fs');
const path = require('path');

function parseM3U(content) {
    const items = [];
    const lines = content.split('\n');
    
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();
        if (line.startsWith('#EXTINF')) {
            // Extract title
            const titleMatch = line.match(/,(.+)$/);
            const title = titleMatch ? titleMatch[1].trim() : 'Unknown';
            
            // Extract thumbnail
            const thumbMatch = line.match(/tvg-logo="([^"]+)"/);
            const thumb = thumbMatch ? thumbMatch[1] : '';
            
            // Extract group
            const groupMatch = line.match(/group-title="([^"]+)"/);
            const group = groupMatch ? groupMatch[1] : 'Channel';
            
            // Next line is URL
            if (i + 1 < lines.length) {
                const url = lines[i + 1].trim();
                if (url && !url.startsWith('#') && url.startsWith('http')) {
                    items.push({
                        title,
                        url,
                        group,
                        thumb,
                        epg: 'Live Stream'
                    });
                }
            }
        }
    }
    
    return items;
}

function parseJSON(content) {
    const items = [];
    try {
        const data = JSON.parse(content);
        if (Array.isArray(data)) {
            for (const item of data) {
                if (item && item.url) {
                    items.push({
                        title: item.title || 'Unknown',
                        url: item.url,
                        group: item.group || 'Channel',
                        thumb: item.thumb || '',
                        epg: item.epg || 'Live Stream'
                    });
                }
            }
        }
    } catch (e) {
        // Not valid JSON
    }
    return items;
}

function parseTXT(content) {
    const items = [];
    const lines = content.split('\n');
    
    for (const line of lines) {
        const trimmed = line.trim();
        if (trimmed && trimmed.startsWith('http')) {
            items.push({
                title: `Channel ${items.length + 1}`,
                url: trimmed,
                group: 'Channel',
                thumb: '',
                epg: 'Live Stream'
            });
        }
    }
    
    return items;
}

function parseJS(content) {
    const items = [];
    const match = content.match(/\[(.*?)\]/s);
    
    if (match) {
        try {
            const data = JSON.parse('[' + match[1] + ']');
            for (const item of data) {
                if (item && item.url) {
                    items.push({
                        title: item.title || 'Unknown',
                        url: item.url,
                        group: item.group || 'Channel',
                        thumb: item.thumb || '',
                        epg: item.epg || 'Live Stream'
                    });
                }
            }
        } catch (e) {
            // Not valid
        }
    }
    
    return items;
}

function main() {
    if (process.argv.length < 3) {
        console.log('\nUsage: node playlist-tool.js <file.m3u|.txt|.json|.js>');
        console.log('\nExample:');
        console.log('  node playlist-tool.js myplaylist.m3u');
        process.exit(1);
    }
    
    const filepath = path.resolve(process.argv[2]);
    
    // Check file exists
    if (!fs.existsSync(filepath)) {
        console.log(`\n❌ Error: File not found: ${filepath}`);
        process.exit(1);
    }
    
    console.log('\n' + '='.repeat(80));
    console.log('🎬 NEXUS TV OS - PLAYLIST EXTRACTION TOOL');
    console.log('='.repeat(80));
    console.log(`File: ${filepath}`);
    console.log(`Size: ${fs.statSync(filepath).size} bytes`);
    console.log();
    
    // Read file
    const content = fs.readFileSync(filepath, 'utf-8');
    
    // Detect format and parse
    const ext = path.extname(filepath).toLowerCase();
    let items = [];
    
    if (['.m3u', '.m3u8'].includes(ext) || content.includes('#EXTINF')) {
        console.log('Format: M3U Playlist');
        items = parseM3U(content);
    } else if (ext === '.json' || content.trim().startsWith('[')) {
        console.log('Format: JSON');
        items = parseJSON(content);
    } else if (ext === '.txt') {
        console.log('Format: Text (one URL per line)');
        items = parseTXT(content);
    } else if (ext === '.js') {
        console.log('Format: JavaScript');
        items = parseJS(content);
    } else {
        console.log('Unknown format. Trying JSON...');
        items = parseJSON(content);
    }
    
    if (items.length === 0) {
        console.log('❌ No valid items found!');
        process.exit(1);
    }
    
    console.log(`✓ Extracted ${items.length} items\n`);
    
    // Show extracted data
    console.log('='.repeat(80));
    console.log('EXTRACTED DATA (showing first 20 items):');
    console.log('='.repeat(80));
    console.log();
    
    for (let i = 0; i < Math.min(20, items.length); i++) {
        const item = items[i];
        console.log(`${i + 1}. TITLE: ${item.title}`);
        console.log(`   URL: ${item.url}`);
        if (item.group) console.log(`   GROUP: ${item.group}`);
        if (item.thumb) console.log(`   THUMB: ${item.thumb}`);
        console.log();
    }
    
    if (items.length > 20) {
        console.log(`... and ${items.length - 20} more items\n`);
    }
    
    // Create output directory
    const outputDir = path.resolve('output');
    const chunksDir = path.join(outputDir, 'chunks');
    
    if (!fs.existsSync(chunksDir)) {
        fs.mkdirSync(chunksDir, { recursive: true });
    }
    
    // Split into chunks
    const numChunks = Math.max(1, Math.min(5, Math.ceil(items.length / 100)));
    const itemsPerChunk = Math.ceil(items.length / numChunks);
    
    console.log('='.repeat(80));
    console.log(`CREATING ${numChunks} CHUNK FILE(S)`);
    console.log('='.repeat(80));
    console.log();
    
    for (let i = 0; i < numChunks; i++) {
        const start = i * itemsPerChunk;
        const end = Math.min(start + itemsPerChunk, items.length);
        const chunkItems = items.slice(start, end);
        
        const chunkName = `chunk_${String(i + 1).padStart(2, '0')}`;
        const jsonFile = path.join(chunksDir, `${chunkName}.json`);
        
        // Write JSON file
        fs.writeFileSync(jsonFile, JSON.stringify(chunkItems, null, 2), 'utf-8');
        
        const stats = fs.statSync(jsonFile);
        console.log(`✓ Created: ${jsonFile}`);
        console.log(`  Items: ${chunkItems.length}`);
        console.log(`  Size: ${stats.size} bytes`);
        console.log();
    }
    
    console.log('='.repeat(80));
    console.log('✅ COMPLETE');
    console.log('='.repeat(80));
    console.log();
    console.log(`Output location: ${chunksDir}`);
    console.log();
    console.log('Files created:');
    
    const files = fs.readdirSync(chunksDir).filter(f => f.endsWith('.json')).sort();
    for (const file of files) {
        const data = JSON.parse(fs.readFileSync(path.join(chunksDir, file), 'utf-8'));
        console.log(`  ✓ ${file} (${data.length} items)`);
    }
    console.log();
}

main();

File: playlist-tool.py
#!/usr/bin/env python3
"""
Nexus TV OS - Command Line Playlist Extraction Tool
Simple, no GUI, shows EVERYTHING in console
"""

import json
import re
import sys
from pathlib import Path

def parse_m3u(content):
    """Parse M3U playlist"""
    items = []
    lines = content.split('\n')
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        if line.startswith('#EXTINF'):
            # Extract title
            title_match = re.search(r',(.+)$', line)
            title = title_match.group(1).strip() if title_match else "Unknown"
            
            # Extract thumbnail
            thumb_match = re.search(r'tvg-logo="([^"]+)"', line)
            thumb = thumb_match.group(1) if thumb_match else ""
            
            # Extract group
            group_match = re.search(r'group-title="([^"]+)"', line)
            group = group_match.group(1) if group_match else "Channel"
            
            # Next line is URL
            if i + 1 < len(lines):
                url = lines[i + 1].strip()
                if url and not url.startswith('#') and url.startswith('http'):
                    items.append({
                        'title': title,
                        'url': url,
                        'group': group,
                        'thumb': thumb,
                        'epg': 'Live Stream'
                    })
            i += 2
        else:
            i += 1
    return items

def parse_json(content):
    """Parse JSON playlist"""
    items = []
    try:
        data = json.loads(content)
        if isinstance(data, list):
            for item in data:
                if isinstance(item, dict) and 'url' in item:
                    items.append({
                        'title': item.get('title', 'Unknown'),
                        'url': item['url'],
                        'group': item.get('group', 'Channel'),
                        'thumb': item.get('thumb', ''),
                        'epg': item.get('epg', 'Live Stream')
                    })
    except:
        pass
    return items

def parse_txt(content):
    """Parse TXT playlist (one URL per line)"""
    items = []
    for line in content.split('\n'):
        line = line.strip()
        if line and line.startswith('http'):
            items.append({
                'title': f"Channel {len(items)+1}",
                'url': line,
                'group': 'Channel',
                'thumb': '',
                'epg': 'Live Stream'
            })
    return items

def parse_js(content):
    """Parse JavaScript playlist"""
    items = []
    match = re.search(r'\[(.*?)\]', content, re.DOTALL)
    if match:
        try:
            data = json.loads('[' + match.group(1) + ']')
            for item in data:
                if isinstance(item, dict) and 'url' in item:
                    items.append({
                        'title': item.get('title', 'Unknown'),
                        'url': item['url'],
                        'group': item.get('group', 'Channel'),
                        'thumb': item.get('thumb', ''),
                        'epg': item.get('epg', 'Live Stream')
                    })
        except:
            pass
    return items

def main():
    if len(sys.argv) < 2:
        print("Usage: python playlist-tool.py <file.m3u|.txt|.json|.js>")
        print("\nExample:")
        print("  python playlist-tool.py myplaylist.m3u")
        sys.exit(1)
    
    filepath = Path(sys.argv[1])
    
    if not filepath.exists():
        print(f"❌ Error: File not found: {filepath}")
        sys.exit(1)
    
    print("\n" + "="*80)
    print("🎬 NEXUS TV OS - PLAYLIST EXTRACTION TOOL")
    print("="*80)
    print(f"File: {filepath.absolute()}")
    print(f"Size: {filepath.stat().st_size} bytes")
    print()
    
    # Read file
    with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
        content = f.read()
    
    # Detect format and parse
    ext = filepath.suffix.lower()
    items = []
    
    if ext in ['.m3u', '.m3u8'] or '#EXTINF' in content:
        print("Format: M3U Playlist")
        items = parse_m3u(content)
    elif ext == '.json' or content.strip().startswith('['):
        print("Format: JSON")
        items = parse_json(content)
    elif ext == '.txt':
        print("Format: Text (one URL per line)")
        items = parse_txt(content)
    elif ext == '.js':
        print("Format: JavaScript")
        items = parse_js(content)
    else:
        print("Unknown format. Trying JSON...")
        items = parse_json(content)
    
    if not items:
        print("❌ No valid items found!")
        sys.exit(1)
    
    print(f"✓ Extracted {len(items)} items\n")
    
    # Show extracted data
    print("="*80)
    print("EXTRACTED DATA (showing first 20 items):")
    print("="*80)
    print()
    
    for i, item in enumerate(items[:20], 1):
        print(f"{i}. TITLE: {item['title']}")
        print(f"   URL: {item['url']}")
        if item['group']:
            print(f"   GROUP: {item['group']}")
        if item['thumb']:
            print(f"   THUMB: {item['thumb']}")
        print()
    
    if len(items) > 20:
        print(f"... and {len(items) - 20} more items\n")
    
    # Create output directory
    output_dir = Path("output").absolute()
    chunks_dir = output_dir / "chunks"
    chunks_dir.mkdir(parents=True, exist_ok=True)
    
    # Split into chunks
    num_chunks = max(1, min(5, len(items) // 100))
    items_per_chunk = len(items) // num_chunks
    
    print("="*80)
    print(f"CREATING {num_chunks} CHUNK FILE(S)")
    print("="*80)
    print()
    
    for i in range(num_chunks):
        start = i * items_per_chunk
        end = start + items_per_chunk if i < num_chunks - 1 else len(items)
        chunk_items = items[start:end]
        
        # Save as JSON
        chunk_name = f"chunk_{i+1:02d}"
        json_file = chunks_dir / f"{chunk_name}.json"
        
        with open(json_file, 'w', encoding='utf-8') as f:
            json.dump(chunk_items, f, indent=2, ensure_ascii=False)
        
        print(f"✓ Created: {json_file}")
        print(f"  Items: {len(chunk_items)}")
        print(f"  Size: {json_file.stat().st_size} bytes")
        print()
    
    print("="*80)
    print("✅ COMPLETE")
    print("="*80)
    print()
    print(f"Output location: {chunks_dir}")
    print()
    print("Files created:")
    for f in sorted(chunks_dir.glob("*.json")):
        with open(f) as file:
            data = json.load(file)
        print(f"  ✓ {f.name} ({len(data)} items)")
    print()

if __name__ == "__main__":
    main()
